Code/GraphMol/ChemReactions/Enumerate/testEnumerate.cpp

```diff
@@ -0,0 +1,243 @@
+#include <RDGeneral/BoostEndInclude.h>
+
+using namespace RDKit;
+
+void pickleTest(EnumerationStrategyBase &en, size_t len) {
+  boost::shared_ptr<EnumerationStrategyBase> base(en.copy());
+  TEST_ASSERT(std::string(base->type()) == std::string(en.type()));
+
+  for (size_t i = 0; i < len; ++i) {
+    std::stringstream ss;
+    {
+      boost::archive::text_oarchive ar(ss);
+      ar &base;
+    }
+    boost::shared_ptr<EnumerationStrategyBase> copy;
+    {
+      boost::archive::text_iarchive ar(ss);
+      ar &copy;
+    }
+    TEST_ASSERT(std::string(base->type()) == std::string(copy->type()));
+    TEST_ASSERT(base->next() == copy->next());
+    TEST_ASSERT(base->getPosition() == en.next());
+  }
+}
+
+void testSamplers() {
+  EnumerationTypes::BBS bbs;
+  bbs.resize(3);
+  for (int i = 0; i < 10; ++i)
+    bbs[0].push_back(boost::shared_ptr<ROMol>(SmilesToMol("C=CCN=C=S")));
+
+  for (int i = 0; i < 5; ++i)
+    bbs[1].push_back(boost::shared_ptr<ROMol>(SmilesToMol("NCc1ncc(Cl)cc1Br")));
+
+  for (int i = 0; i < 6; ++i)
+    bbs[2].push_back(
+        boost::shared_ptr<ROMol>(SmilesToMol("NCCCc1ncc(Cl)cc1Br")));
+
+  ChemicalReaction rxn;
+  CartesianProductStrategy cart;
+  cart.initialize(rxn, bbs);
+  RandomSampleStrategy rand;
+  rand.initialize(rxn, bbs);
+  RandomSampleAllBBsStrategy randBBs;
+  randBBs.initialize(rxn, bbs);
+  EvenSamplePairsStrategy even;
+  even.initialize(rxn, bbs);
+  std::vector<boost::shared_ptr<EnumerationStrategyBase> > enumerators;
+  enumerators.push_back(
+      boost::shared_ptr<EnumerationStrategyBase>(cart.copy()));
+  enumerators.push_back(
+      boost::shared_ptr<EnumerationStrategyBase>(rand.copy()));
+  enumerators.push_back(
+      boost::shared_ptr<EnumerationStrategyBase>(randBBs.copy()));
+  enumerators.push_back(
+      boost::shared_ptr<EnumerationStrategyBase>(even.copy()));
+
+  for (size_t i = 0; i < enumerators.size(); ++i) {
+    TEST_ASSERT(enumerators[i]->getNumPermutations() == 10 * 5 * 6);
+    pickleTest(*enumerators[i], 10 * 5 * 6);
+  }
+
+  // for(auto&& i: enumerators) {
+  //  TEST_ASSERT(i->getNumPermutations() == 10*5*6);
+  //}
+}
+
+void testEvenSamplers() {
+  EnumerationTypes::BBS bbs;
+  bbs.resize(3);
+  unsigned long R1 = 6000;
+  unsigned long R2 = 500;
+  unsigned long R3 = 10000;
+  for (unsigned long i = 0; i < R1; ++i)
+    bbs[0].push_back(boost::shared_ptr<ROMol>(SmilesToMol("C=CCN=C=S")));
+
+  for (unsigned long i = 0; i < R2; ++i)
+    bbs[1].push_back(boost::shared_ptr<ROMol>(SmilesToMol("NCc1ncc(Cl)cc1Br")));
+
+  for (unsigned long i = 0; i < R3; ++i)
+    bbs[2].push_back(
+        boost::shared_ptr<ROMol>(SmilesToMol("NCCCc1ncc(Cl)cc1Br")));
+
+  ChemicalReaction rxn;
+  EvenSamplePairsStrategy even;
+  even.initialize(rxn, bbs);
+  std::cout << even.getNumPermutations() << " " << R1 * R2 * R3 << std::endl;
+  TEST_ASSERT(even.getNumPermutations() == R1 * R2 * R3);
+
+  for (size_t i = 0; i < 5000; ++i) {
+    even.next();
+  }
+  even.stats();
```

Comment:
It looks like the only thing you're actually testing in this function is the size of the enumerator. Shouldn't you also have a test in here for the results of the sampling?
I guess this would require being able to seed the RNG in the strategy, but that's probably not a bad thing.
(Ok, this is the `EvenSamplePairsStrategy`, which doesn't actually have an RNG, but the point holds for the other strategies that do)